pyocd/probe/picoprobe.py

```diff
@@ -0,0 +1,636 @@
+            raise ValueError("unsupported wire protocol %s" % protocol)
+
+        self._is_connected = True
+        # Use the bulk or safe read and write functions according to option
+        if self.session.options.get(self.SAFESWD_OPTION):
+            self.read_ap_multiple = self._safe_read_ap_multiple
+            self.write_ap_multiple = self._safe_write_ap_multiple
+        else:
+            self.read_ap_multiple = self._bulk_read_ap_multiple
+            self.write_ap_multiple = self._bulk_write_ap_multiple
+        # Subscribe to option change events
+        self.session.options.subscribe(self._change_options, [self.SAFESWD_OPTION])
+        # Do I need to do anything else here?
+        # SWJ switch sequence is handled externally...
+
+    def swj_sequence(self, length, bits):
+        self._link.start_queue()
+        self._link.q_write_bits(bits, length)
+        self._link.flush_queue()
+
+    def disconnect(self):
+        self._is_connected = False
+        self._select = -1
+
+    def set_clock(self, frequency):
+        self._link.set_swd_frequency(frequency // 1000)
+
+    def reset(self):
+        self.assert_reset(True)
+        sleep(self.session.options.get('reset.hold_time'))
+        self.assert_reset(False)
+        sleep(self.session.options.get('reset.post_delay'))
+
+    def assert_reset(self, asserted):
+        self._select = -1
+        self._link.assert_target_reset(asserted)
+        self._reset = asserted
+
+    def is_reset_asserted(self):
+        # No support for reading back the current state
+        return self._reset
+
+    # ------------------------------------------- #
+    #          DAP Access functions
+    # ------------------------------------------- #
+    def read_dp(self, addr, now=True):
+        val = self._read_reg(addr, self.DP)
+
+        # Return the result or the result callback for deferred reads
+        def read_dp_result_callback():
+
+            return val
+        return val if now else read_dp_result_callback
+
+    def write_dp(self, addr, value):
+        # Check whether we need to write a new SELECT value
+        if addr == self.SELECT:
+            if value == self._select:
+                return
+            else:
+                self._select = value
+        self._write_reg(addr, self.DP, value)
+
+    def read_ap(self, addr, now=True):
+        (ret,) = self.read_ap_multiple(addr)
+
+        def read_ap_cb():
+            return ret
+        return ret if now else read_ap_cb
+
+    def write_ap(self, addr, value):
+        self.write_ap_multiple(addr, (value,))
+
+    def _safe_read_ap_multiple(self, addr, count=1, now=True):
+        # Extract the APSEL and APBANKSEL fields
+        sel = addr & self.APSEL_APBANKSEL
+        # Write it to SELECT register on DP
+        # It would appear that a write to SELECT is always done first,
+        # so this is possibly not needed. Oh well, it's cached anyway.
+        self.write_dp(self.SELECT, sel)
```

Comment:
Reminder that you can remove `SELECT` management. Same for other AP methods, of course.